staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/subresource.go

```diff
@@ -0,0 +1,192 @@
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+)
+
+var (
+	scaleGroupVersion = schema.GroupVersion{Group: "autoscaling", Version: "v1"}
+	scalePath         = fieldpath.MakePathOrDie("spec", "replicas")
+)
+
+// SubresourceScaler manages the conversion of managed fields between a main
+// resource and subresources
+type SubresourceScaler struct {
+	parentEntries      []metav1.ManagedFieldsEntry
+	parentGroupVersion schema.GroupVersion
+	parentPath         fieldpath.Path
+}
+
+// ScalePathMappings maps a group/version to the replicas path in a scale subresource
+type ScalePathMappings map[schema.GroupVersion]fieldpath.Path
+
+// NewSubresourceScaler creates a new SubresourceScaler
+func NewSubresourceScaler(parentEntries []metav1.ManagedFieldsEntry, parentGroupVersion schema.GroupVersion, parentPath fieldpath.Path) (*SubresourceScaler, error) {
+	return &SubresourceScaler{
+		parentEntries:      parentEntries,
+		parentGroupVersion: parentGroupVersion,
+		parentPath:         parentPath,
+	}, nil
+}
+
+// ToSubresource filter the managed fields of the main resource and convert
+// them so that they can be handled by a subresource.
+// For the managed fields that have a path that could be handled by a
+// subresource this method performs two changes:
+// 1. The APIVersion is changed to the APIVersion of the subresource
+// 2. The Fields are transformed to the path of that subresource
+func (h *SubresourceScaler) ToSubresource() ([]metav1.ManagedFieldsEntry, error) {
+	managed, err := internal.DecodeObjectManagedFields(h.parentEntries)
+	if err != nil {
+		return nil, err
+	}
+
+	f := fieldpath.ManagedFields{}
+	t := map[string]*metav1.Time{}
+	for manager, versionedSet := range managed.Fields() {
+
+		if versionedSet.Set().Has(h.parentPath) {
+			newVersionedSet := fieldpath.NewVersionedSet(
+				fieldpath.NewSet(scalePath),
+				fieldpath.APIVersion(scaleGroupVersion.String()),
+				versionedSet.Applied(),
+			)
+
+			f[manager] = newVersionedSet
+			if operationTime, ok := managed.Times()[manager]; ok {
+				t[manager] = operationTime
+			}
+		}
+
+	}
+
+	return managedFieldsEntries(internal.NewManaged(f, t))
+}
+
+// ToParent merges the `subresourceEntries` with the parent entries and
+// transform them accordingly
+func (h *SubresourceScaler) ToParent(scaleEntries []metav1.ManagedFieldsEntry) ([]metav1.ManagedFieldsEntry, error) {
+	f := fieldpath.ManagedFields{}
+	t := map[string]*metav1.Time{}
+
+	// 1. Decode the scale entry and change it back to the parent API version and
+	// the parent path
+	scaleManagedFields, err := internal.DecodeObjectManagedFields(scaleEntries)
+	if err != nil {
+		return nil, err
+	}
+
+	for manager, versionedSet := range scaleManagedFields.Fields() {
+
+		if versionedSet.Set().Has(scalePath) {
+			// TODO (nodo): handle multiple managed fields versions.
+			newAPIVersion := h.parentGroupVersion.String()
+			newVersionedSet := fieldpath.NewVersionedSet(
```

Comment:
Isn't that replacing all the fieldsets that had the replicas field before? I don't think that's a good enough reason to remove it.

Reply:
I think with the new algorithm this is resolved.